crates/mbe/src/expander/matcher.rs

```diff
@@ -710,6 +710,7 @@ fn match_meta_var(kind: &str, input: &mut TtIter) -> ExpandResult<Option<Fragmen
             let tt_result = match kind {
                 "ident" => input
                     .expect_ident()
+                    .and_then(|ident| if ident.text == "_" { Err(()) } else { Ok(ident) })
```

Comment:
Could it be moved inside `expect_ident` ?

Reply:
Some of the other callers of `expect_ident` do want to allow `_`:
  * `expect_lifetime`: previously `$x:lifetime` doesn't accept `'_` but should to match rustc (see added `test_underscore_lifetime` test)
  * `expect_separator`: should accept `_` for `$(x)_*` (did previously as `Punct`)
  * `eat_fragment_kind`: doesn't need to accept `_`, but it would just get rejected later in `match_meta_var`

Reply:
Oh I see. than maybe add a comment for that ? Something like the following is good enough for me :

```
// Special case for rejecting underscore as ident in matcher :
// See https://github.com/rust-lang/rust/pull/48842
```

Reply:
Actually, for `eat_fragment_kind`, accepting `_` as in `$x:_` moves the error from the macro definition to the macro call site.

What do you think of making `expect_ident` disallow `_` and adding `expect_ident_or_underscore` for use in `expect_lifetime` and `expect_separator`?

Reply:
Sounds good to me !